designs/drift-hash-versioning.md

```diff
@@ -0,0 +1,162 @@
+  annotations: 
+    karpenter.k8s.aws/ec2nodeclass-hash: mnopqr-stuvwx
+    ...
+---
+apiVersion: karpenter.sh/v1beta1
+kind: Nodeclaim 
+metadata:
+  name: default
+  annotations: 
+    karpenter.sh/nodepool-hash: abcdef-ghijkl
+    karpenter.k8s.aws/ec2nodeclass-hash: mnopqr-stuvwx
+    ...
+```
+
+As the Karpenter CRDs change, Karpenter’s underlying hashing function may also change. To handle the upgrade path, Karpenter will prepend a hash of the CRD to the customer resource spec hash. Karpenter will assert that this struct version is equivalent before checking the customer resource spec hash. 
+
+As a result, two annotations will be used analyze and orchestrate drift: 
+
+* nodepool-hash: (nodepool-crd-version)-(nodepool-hash)
+* ec2nodeclass-hash: (ec2nodeclass-crd-version)-(ec2nodeclass-hash)
+
+### Option 1(Recommended): Fallback to rehashing on the NodeClaim 
+
+Option 1 will have two case for drift:
+
+1. When the CRD `nodepool-crd-version/ec2nodeclass-crd-version` matches
+    1. Drift will work as normal and the `NodePool/EC2NodeClass` hash will be compared to evaluate for drift 
+2. When the CRD `nodepool-crd-version/ec2nodeclass-crd-version` does not match
+    1. Karpenter would recalculate the hash of the `Nodepool/EC2NodeClass` and back-propagate the annotation to the NodeClaims
+    2. Karpenter will need to lock drift on `NodeClaims` that were considered drifted prior to the Karpenter upgrade. 
+
+Pros/Cons
+
+- 👍👍 Simple implementation effort 
+- 👍 Cover general cases of updates to the NodePool/EC2NodeClass CRD or the hashing mechanism
+- 👎 Nodes that were considered drifted before the Karpenter upgrade are not able to be un-drifted after a Karpenter upgrade 
+- 👎 Does not handle cases where users change NodePool/EC2NodeClass resources as Karpenter is updating
+
+### Option 2: Fallback to individual field evaluation for drift
+
+Similar to Option 1, there will be two cases for drift:
+
+1. When the `nodepool-crd-version/ec2nodeclass-crd-version` matches
+    1. Drift will work as normal and the `NodePool/EC2NodeClass` hash will be compared to evaluate for drift 
+2. When the `nodepool-crd-version/ec2nodeclass-crd-version` does not match
+    1. Karpenter will individually check each field for drift. *This would be an adaptation of the (Deployment Controller)[https://github.com/kubernetes/kubernetes/blob/83e6636096548f4b49a94ece6bc4a26fd67e4449/pkg/controller/deployment/util/deployment_util.go#L603] for podTemplateSpec relationship to pods.*
+    2. Karpenter will need to populate the `NodeClaim` with values that were used in launching the instances. Drift evaluation will need a point of reference on manual checking.
+
+Pros/Cons
+
+- 👍 Cover general cases of updates to the `NodePool/EC2NodeClass` CRD or the hashing mechanism 
+- 👍 Handles cases where users change `NodePool/EC2NodeClass` resources as Karpenter is updating
+- 👎👎 A lot of implementation and maintenance effort required in supporting individual field evaluation 
```

Comment:
I would cautious against this as the main reason to not do this. I get this is problematic to maintain but if this solution is more accurate or correct there is value in doing this. 

Its worth explaining if the other solution doesn't cover all the bases why those may not matter or address them in the solution. You should add to what is the risk of having the burden of maintaining this behavior.

Reply:
Defiantly agree here, the main concern is a performance of karpenter when karpenter has to do the individual field evaluation for large clusters.